tempodb/blocklist/poller.go

```diff
@@ -226,69 +256,141 @@ func (p *Poller) pollTenantAndCreateIndex(ctx context.Context, tenantID string)
 	return blocklist, compactedBlocklist, nil
 }
 
-func (p *Poller) pollTenantBlocks(ctx context.Context, tenantID string) ([]*backend.BlockMeta, []*backend.CompactedBlockMeta, error) {
-	blockIDs, err := p.reader.Blocks(ctx, tenantID)
+func (p *Poller) pollTenantBlocks(
+	ctx context.Context,
+	tenantID string,
+	previous *List,
+) ([]*backend.BlockMeta, []*backend.CompactedBlockMeta, error) {
+	span, derivedCtx := opentracing.StartSpanFromContext(ctx, "Poller.pollTenantBlocks")
+	defer span.Finish()
+
+	currentBlockIDs, currentCompactedBlockIDs, err := p.reader.Blocks(derivedCtx, tenantID)
 	if err != nil {
-		metricBlocklistErrors.WithLabelValues(tenantID).Inc()
-		return []*backend.BlockMeta{}, []*backend.CompactedBlockMeta{}, err
+		return nil, nil, err
 	}
 
-	bg := boundedwaitgroup.New(p.cfg.PollConcurrency)
-	chMeta := make(chan *backend.BlockMeta, len(blockIDs))
-	chCompactedMeta := make(chan *backend.CompactedBlockMeta, len(blockIDs))
+	readPreviousSpan, _ := opentracing.StartSpanFromContext(derivedCtx, "readPreviousResults")
+	metas := previous.Metas(tenantID)
+	readPreviousSpan.SetTag("metas", len(metas))
+	compactedMetas := previous.CompactedMetas(tenantID)
+	readPreviousSpan.SetTag("compactedMetas", len(compactedMetas))
+	readPreviousSpan.Finish()
+
+	assembleSpan, _ := opentracing.StartSpanFromContext(derivedCtx, "assemblePreviousResults")
+	mm := make(map[uuid.UUID]*backend.BlockMeta, len(metas))
+	cm := make(map[uuid.UUID]*backend.CompactedBlockMeta, len(compactedMetas))
+
+	for _, i := range metas {
+		mm[i.BlockID] = i
+	}
+
+	for _, i := range compactedMetas {
+		cm[i.BlockID] = i
+	}
+	assembleSpan.Finish()
+
+	mtx := sync.Mutex{}
+	newBlockList := make([]*backend.BlockMeta, 0, len(currentBlockIDs))
+	newCompactedBlocklist := make([]*backend.CompactedBlockMeta, 0, len(currentCompactedBlockIDs))
 	anyError := atomic.Error{}
 
-	for _, blockID := range blockIDs {
+	learnSpan, _ := opentracing.StartSpanFromContext(derivedCtx, "learnFromPrevious")
+	newBlockIDs := []uuid.UUID{}
+	for _, blockID := range currentBlockIDs {
```

Comment:
we know which uuids are compacted and which are not. we lose that by putting everything in a newBlockIDs slice and then calling pollBlock. ideally we would use this information to reduce backend calls and have more direct? purposeful? terse? code. 

however, this is simple and easy to follow so i'm fine with it. it's still a very large improvement over current.

Reply:
The `newBlockIDs` slice is used to build a list of UUIDs that we have not previously fetched, and then poll them.  What information are we losing here by building the slice?  We still need to fetch the meta in order to build up the blocklist.

Reply:
At this point we are aware which block ids are compacted and which are not, but we put those all in the same slice and lose that information.

The "pollBlock" function pulls the meta and if that fails it bulls the compacted meta. We still need this 2 tiered approach for the metas, but for the compacted metas we can skip the first GET and go straight to trying to grab the compacted meta.

Reply:
I saw that also, but didn't chase it down.  I've added a TODO for this.